core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/PushPredicateIntoTableScan.java

```diff
@@ -233,8 +235,9 @@ public static Optional<PlanNode> pushFilterIntoTableScan(
             return Optional.of(node);
         }
 
-        if (newDomain.isNone()) {
-            // TODO: DomainTranslator.fromPredicate can infer that the expression is "false" in some cases (TupleDomain.none()).
+        if (newDomain.isNone() ||
+                (decomposedPredicate.getRemainingExpression() instanceof Cast cast && cast.getExpression() instanceof NullLiteral)) {
+            // TODO: DomainTranslator.getExtractionResult can infer that the expression is "false" (TupleDomain.none()) or "null" in some cases.
```

Comment:
The reason why you're adding a check here is that such Filter wasn't removed by some other rule.
What if a null-filter eligible for removal is not directly above a Table Scan?
My understanding is that it won't get removed, but it should.

I think we need a check here, in PushPredicateIntoTableScan, but we definitely need a check elsewhere as well.

Reply:
I agree, but I don't think there's a way to come up with a rule that can cover 100% of the cases because:
1. AFAIU we don't want to evaluate the entire expression within the rule (source: https://github.com/trinodb/trino/pull/15558#discussion_r1059074027).
2. Comparing (except `IS DISTINCT FROM`) `null` to anything should result in `null`, not `false` (source: https://trino.io/docs/current/functions/comparison.html#is-distinct-from-and-is-not-distinct-from). I can modify `io.trino.sql.planner.ExpressionInterpreter.Visitor#processComparisonExpression` to return `false` instead of `null` when `left` \ `right` is `null`, but AFAIU this is not logically correct (BTW I've tried this and got lots of tests failures).

I'm not sure I agree with the comment that existed even before my change: https://github.com/trinodb/trino/blob/208b38060b5207a57209dcc98aec89eb9240a86d/core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/PushPredicateIntoTableScan.java#L237-L239 Doing the same logic twice just to spare this `if`? And what if the logic of `PushPredicateIntoTableScan` will be changed in the future? If we can't come up with a rule that can cover 100% of the filters that are eventually evaluated to `false`/ `null`, then IMHO, we should rely on this `if`. WDYT?

Reply:
> We're not evaluating here either, right?

Right, we're executing some logic that's specific for the pushdown and discovering `false` filters on the way.

> What prevents us from having a Rule that would e.g. apply to every FilterNode and did the logic which we're doing here?

It feels arbitrary to me to execute some logic that we happen to have that's doing part of the job, but that must be a minority opinion since both you and the comment writer think otherwise.

> I think we need a check here, in PushPredicateIntoTableScan, but we definitely need a check elsewhere as well.

So I'll leave this `if` as is and add another rule

Reply:
As in https://github.com/trinodb/trino/pull/16206#discussion_r1118616081, i think we need both

- a separate rule, which would cover for Filter-not-necessarily-right-above-a-TableScan that is redundant and can be removed
- the change here as proposed is also fine to have. Rules should avoid depending on each other as much as possible. However, we should strive so that reverting the change here in PushPredicateIntoTableScan.java would not impact what query plans we produce.